corehq/ex-submodules/pillowtop/reindexer/reindexer.py

```diff
@@ -159,6 +160,13 @@ def prepare_index_for_usage(es, index_info):
     es.indices.refresh(index_info.index)
 
 
+def _override_alias(index_info, es_client):
+    # Override the alias during the reindex so that production
+    #   could use the alias pointing to existing index for live operations.
```

Comment:
@sravfeyn can you talk a little more about this change?  It feels weird to dynamically patch the alias name like this - it'll be hard to tell in code whether the alias has already been changed, and seems like it even runs the risk of changing it twice.  Is it necessary to refer to the index by alias at all?  A reindex does seem like one context where we'd want to refer explicitly to an index by its name.

Reply:
Sure, this is a good concern.

A while ago we started using index aliases instead of direct index names. So most of the code uses index alias to interface with Elasticsearch, including pillow processors which are used by the reindexers.

So, when we want to trigger a reindex, we have to update both the index name and the index alias in the code. What this change is intending to do is that the reindexer auto assigns a particular alias and uses that to do the reindex.

> It feels weird to dynamically patch the alias name like this -

Agreed, I didn't feel very good while writing this. But what I tried is to keep the override method as a private thing that's used just in the broader reindexer method.

Reply:
I think it'll mutate the global `index_info` constant in question, though, won't it?  That shouldn't be an issue as long as the reindex is the only thing to happen in this process, but it seems like it's inviting trouble.

> A while ago we started using index aliases instead of direct index names. So most of the code uses index alias to interface with Elasticsearch, including pillow processors which are used by the reindexers.

Do you have a sense of what it'd look like to change this?  I do think it'd make sense for reindexers to work with names, not aliases, since aliases are dynamic.

Another possible approach would be to create a new `index_info` instance with this modified alias when instantiating the reindexer.  That would ensure this patch is only applied to the reindexer class.

Reply:
> I think it'll mutate the global index_info constant in question, though, won't it?

> Another possible approach would be to create a new index_info instance with this modified alias when instantiating the reindexer. That would ensure this patch is only applied to the reindexer class.

Yes, it does mutate the global variable, but the reindex runs as a management commands for which this was fine. But, copying is definitely much better, so I have modified it per this suggestion. Thanks for the idea!

> Do you have a sense of what it'd look like to change this? I do think it'd make sense for reindexers to work with names, not aliases, since aliases are dynamic.

It was an intentional change to use aliases. But I agree, for reindexers it is perhaps better to use index-names. But it does require passing that flag down various methods (such as reindexer factories, processors, individual pillow processors and finally ElasticsearchInterface). By copying/overriding the alias all of that is being avoided.

Reply:
> It was an intentional change to use aliases.

Oh yeah, I agree that aliases are better for querying, I meant to change only for the reindexing process.  Looking at the reindexing code now it does seem to mostly use `index_info.index`, with the notable exception of `initialize_index_and_mapping`, which explicitly sets the alias on the index, and `build_bulk_payload`, which writes to an alias.

Thanks for the switch to a copy, as you said, it probably won't matter, but I could see it causing issues if we say, made a util to add domains to the case search index from the UI.  Could you instead make the copy in the constructor, however?  For instance, the `clean` method is called before `reindex`, but presumably it should be working on the same index.

I'm also not a fan of the conditional there - in what circumstances would it evaluate to True and when to False?  What if a reindex fails partway through and is picked back up?  What if that happens the next day?  Could the `clean` method inadvertently wipe the live index?

Would this be a fair substitution?
```python
def __init__(...):
    ...
    self.index_info = ElasticsearchIndexInfo(
        index=index_info.index,
        alias=f"{index_info.index}__reindex",
        ...
    )
```
That way the alias would be determined only by the index name, not what the current state of the alias is or what the date is.

I don't feel like I have my head wrapped around this quite well enough to speak conclusively, just raising concerns I have.  @snopoke would you mind weighing in?

Reply:
Actually, I am leaning towards abandoning this idea as it's quite a fragile approach. Instead, we will just update the alias as well in the mapping (as it is now) when we want to change the mapping.